packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -1988,7 +2023,7 @@ class InputDecoration {
       counterStyle: counterStyle ?? this.counterStyle,
       filled: filled ?? this.filled,
       fillColor: fillColor ?? this.fillColor,
-      border: border ?? this.border,
+      border: this.border is _DefaultInputBorder ? border : this.border,
```

Comment:
this is weird

Reply:
as we discussed in person:

Rather than _DefaultInputBorder, I suggest we require that the border eventually compute to a non-null value (null here meaning "inherit" as with everything else). To get no border, you specify an InputBorder that doesn't paint a border (but does provide a shape). That shape controls where the fill region fills. We can get rid of the code path that tries to avoid widgets when there's no border; that way it's easier to animate between different filled shapes even when there's no border.

Reply:
As a side-effect of this, you may want to move the borderSide computation into InputBorder subclasses, and instead of using copyWith(borderSide), have an apply method that applies a color, isCollapsed, enabled, and isFocused. That way you can remove the hard-coded knowledge about borderSide from InputBorder, since the new "BorderlessInputBorder" (or whatever you call it) won't need a borderSide but may still have opinions about the rest.